libyara/modules/pe.c

```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
       rich_internal(module(), integer_argument(2), integer_argument(1)));
 }
 
+
+define_function(calculate_checksum)
+{
+  YR_OBJECT* module = module();
+
+  PE* pe = (PE*) module->data;
+  if (pe == NULL)
+    return_integer(UNDEFINED);
+
+  uint64_t csum = 0;
+  int csum_offset = ((uint8_t*)&(pe->header->OptionalHeader) +
+		     offsetof(IMAGE_OPTIONAL_HEADER32, CheckSum)) - pe->data;
+  for (int i=0; i <= pe->data_size / 4; i++)
+  {
+    // Treat the CheckSum field as 0 -- the offset is the same for
+    // PE32 and PE64.
+    if (4*i == csum_offset)
+      continue;
+    if (4*i+4 < pe->data_size)
+    {
+      csum += ((uint64_t)pe->data[4*i] +
+	       ((uint64_t)pe->data[4*i+1] << 8)  +
+	       ((uint64_t)pe->data[4*i+2] << 16) +
+	       ((uint64_t)pe->data[4*i+3] << 24));
+    }
+    else
+    {
+      switch (pe->data_size % 4)
+      {
+        case 3: csum += (uint64_t)pe->data[4*i+2] << 16;
+        case 2: csum += (uint64_t)pe->data[4*i+1] << 8;
+        case 1: csum += (uint64_t)pe->data[4*i];
+      }
```

Comment:
Clever use of the switch fall through, but I think a for loop would make the intention more evident.